packages/flutter/lib/src/material/input.dart

```diff
@@ -368,24 +326,159 @@ class _InputState extends State<Input> {
+  // InputContainer, InputFormField, InputField.
+  Input({
+    Key key,
+    this.value,
+    this.keyboardType: TextInputType.text,
+    this.icon,
+    this.labelText,
+    this.hintText,
+    this.errorText,
+    this.style,
+    this.hideText: false,
+    this.isDense: false,
+    this.autofocus: false,
+    this.maxLines: 1,
+    this.onChanged,
+    this.onSubmitted,
+  }) : super(key: key);
+
+  /// The current state of text of the input field. This includes the selected
+  /// text, if any, among other things.
+  final InputValue value;
+
+  /// The type of keyboard to use for editing the text.
+  final TextInputType keyboardType;
+
+  /// An icon to show adjacent to the input field.
+  ///
+  /// The size and color of the icon is configured automatically using an
+  /// [IconTheme] and therefore does not need to be explicitly given in the
+  /// icon widget.
+  ///
+  /// See [Icon], [ImageIcon].
+  final Widget icon;
+
+  /// Text to show above the input field.
+  final String labelText;
+
+  /// Text to show inline in the input field when it would otherwise be empty.
+  final String hintText;
+
+  /// Text to show when the input text is invalid.
+  final String errorText;
+
+  /// The style to use for the text being edited.
+  final TextStyle style;
+
+  /// Whether to hide the text being edited (e.g., for passwords).
+  ///
+  /// When this is set to true, all the characters in the input are replaced by
+  /// U+2022 BULLET characters (•).
+  final bool hideText;
+
+  /// Whether the input field is part of a dense form (i.e., uses less vertical space).
+  final bool isDense;
+
+  /// Whether this input field should focus itself if nothing else is already focused.
+  final bool autofocus;
+
+  /// The maximum number of lines for the text to span, wrapping if necessary.
+  /// If this is 1 (the default), the text will not wrap, but will scroll
+  /// horizontally instead.
+  final int maxLines;
+
+  /// Called when the text being edited changes.
+  ///
+  /// The [value] must be updated each time [onChanged] is invoked.
+  final ValueChanged<InputValue> onChanged;
+
+  /// Called when the user indicates that they are done editing the text in the field.
+  final ValueChanged<InputValue> onSubmitted;
+
+  @override
+  _InputState createState() => new _InputState();
+}
+
+class _InputState extends State<Input> {
+  final GlobalKey<_InputFieldState> _inputFieldKey = new GlobalKey<_InputFieldState>();
+  final GlobalKey _focusKey = new GlobalKey();
+
+  GlobalKey get focusKey => config.key is GlobalKey ? config.key : _focusKey;
+
+  @override
+  Widget build(BuildContext context) {
     return new GestureDetector(
-      key: config.key is GlobalKey ? null : focusKey,
-      behavior: HitTestBehavior.opaque,
+      key: focusKey == _focusKey ? _focusKey : null,
       onTap: () {
         _inputFieldKey.currentState?.requestKeyboard();
       },
-      child: child,
+      child: new Builder(
```

Comment:
I'll add this comment:
```
Since the focusKey may have been created here, defer building the InputContainer until the focusKey's context has been set. This is necessary because we're passing the value of Focus.at(focusKey.currentContext, ...) along to the InputContainer.
```